### ISSUE TITLE
Memory error during `push rbx` leave instruction executed partially

### ISSUE THREAD

**Issue by mrexodia** [icicle-emu/icicle-emu#66]
This issue was initially reported by @huettenhain at https://github.com/icicle-emu/icicle-python/issues/10.

Test code in rust: https://github.com/icicle-emu/icicle-python/blob/aef686faa6a66d73da5a18c7d06c432a18db4388/tests/tests.rs#L139-L156

It appears that this is due to how to semantics are implemented in Ghidra, where the `STACKPTR` is updated first and the memory write happens next:

```
macro push48(x) {
  mysave:8 = x;
  $(STACKPTR) = $(STACKPTR) - 8;
  *:8 $(STACKPTR) = mysave;
}
```

Not sure if there is a generic solution for this. Moving memory writes up would most likely require major work and I am not sure if push/pop are the only affected instructions, but for now we could update the semantics.

**Comment by mchesser** [comment]
Yeah, at some point I look at trying to fix this -- but, particularly for x86-64, there are a huge number of instructions where this is an issue. 

A vast majority of them I think can be fixed with a automatic rewrite rule that move load/stores above register updates -- but I know of at least a few cases where manual changes are required underlying SLEIGH specification.

I _do_ want to implement some automatic rewrites, but I've held off a bit, since I wasn't sure if having a partial solution would make it harder to detect the cases that still have issues. The other cases that need some thinking about include: 

- Cases where the SLEIGH specification (incorrectly) duplicates loads/stores within a single instruction. It should be pretty easy to dedup these, but some care is needed for instructions that loads/store from multiple addresses.
- Large memory access being broken up into multiple Pcode load/store ops. This is most relevant for store operations where memory writes are more difficult to undo. Internally, due to Icicle code, this can happen for unaligned accesses as well. This could probably be fixed by checking that all inner addresses are resolveable before actually updating the state, but this is a little complicated.
- Memory accesses in delay slots.

As far as I known, there are also instructions in some ISAs which do result in partial execution on memory exceptions (e.g., when faults cross page boundaries).

**Comment by mrexodia** [comment]
For now I just worked around the x86-specific issue with `push`, because it seems like a common-enough case to warrant a sleigh update: https://github.com/icicle-emu/icicle-python/commit/37c0fc57a8e286381569688a2d6ab028841ea56e

It would be interesting to write some analysis pass that tries to automatically identify which instructions are affected. The Ghidra people are probably also interested in fixing this in the sleigh spec, so might also be worth bringing it to their attention upstream.

Another (slower) solution could be to 'simply' not commit stores to the actual memory/context until the end of the instruction (or even execution), but this most likely violates a bunch of assumptions and would be tricky to implement.

**Comment by mchesser** [comment]
I thought about splitting memory checks from the final commit, but yeah it's hard to implement efficiently while maintaining support for hooks, and other dynamic manipulations. 

I also want to add support for a 'passthrough' memory mode, where guest addresses are directly translated to host address -- though this would likely have some limitations, so the existing memory mode would still exist.
